cove_360/lib/threesixtygiving.py

```diff
@@ -1,19 +1,30 @@
-import json
+import datetime
 import itertools
-import openpyxl
+import json
+import re
 from collections import OrderedDict, Callable
 from decimal import Decimal
-import datetime
-import pytz
-import re
+
 import libcove.lib.tools as tools
+import openpyxl
+import pytz
+from dateutil.relativedelta import relativedelta
 from django.utils.html import mark_safe
 from libcove.lib.common import common_checks_context, get_orgids_prefixes
 from rangedict import RangeDict as range_dict
 
 QUALITY_TEST_CLASS = 'quality_accuracy'
 USEFULNESS_TEST_CLASS = 'usefulness'
 
+GRANT_DATES = {}
```

Comment:
This mechanism is a great idea that will help to speed up the runtime of this function. This implementation doesn't feel super Pythonic, though. I've done a quick bit of Googling and it looks like there's a range of memoization (I learned a new word today!) decorators that achieve the same ends but are more Pythonic and I think also make this code a bit less fragile. 

Cc-ing some actual Python devs (@kindly ? @Bjwebb ?) for their view, though!

Reply:
Yeah, I'm generally in favour of memoisation decorators, although I've not had chance to review what this does exactly.

Reply:
I didn't know about `memoisation decorators`. I will look into it. Thanks @robredpath!

@Bjwebb there are a few checks for dates. Instead of going through the dates and format them in each check, it is done once at the beginning and store the format dates and error messages in this dictionary. Then in each check the date or error message is check directly in the dict.

Reply:
@robredpath @Bjwebb I have been looking at memoisation and I think it might not make the function much more efficient but correct me if I'm wrong.

Reply:
@BibianaC as I understand it, you've basically re-implemented memoization here, but in a way that's a bit more fragile. I don't think that 'real' memoization would actually be any faster, because it would be doing basically the same thing. 

The benefit of doing it using a memoization decorator would be that the code would be a bit easier to read (global-scope dicts are fairly uncommon in Python, so it's a bit jarring to see) and it would make the code easier to refactor in the future, if we wanted to. 

It might also save on some memory - for a very large file, the dict could get pretty big. It looks like [`lru_cache'](https://docs.python.org/3/library/functools.html#functools.lru_cache) lets you limit the size of the cache. 

To be clear - I'm suggesting:
* Refactoring `create_grant_dates_dict` to operate on a single grant each time 
* Calling that function instead of accessing the dict in each of the additional checks that uses the data

That said, this isn't a blocker for me; just a suggestion! I'd be perfectly happy to keep this as it is, and to make a note of a potential improvement for the future.

Reply:
@robredpath I have implemented it without a decorator though. 
I have disabled the max size of the cache but I can modify it. The default one is 128.